src/tls.py

```diff
@@ -0,0 +1,260 @@
+        )
+        self.framework.observe(
+            self.charm.on[TLS_RELATION].relation_joined, self._on_certificates_joined
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_available, self._on_certificate_available
+        )
+        self.framework.observe(self.charm.on.set_tls_private_key_action, self._set_tls_private_key)
+
+    def _on_certificates_created(self, _):
+        """"""
+        if not self.charm.unit.is_leader():
+            return
+
+        self.peer_relation.data[self.charm.app].update({"tls": "enabled"})
+
+    def _on_certificates_joined(self, event: RelationJoinedEvent) -> None:
+        """"""
+        # generate unit private key if not already created by action
+        if not self.private_key:
+            self.peer_relation.data[self.charm.unit].update(
+                {"private-key": generate_private_key().decode("utf-8")}
+            )
+
+        # generate unit private key if not already created by action
+        if not self.keystore_password:
+            self.peer_relation.data[self.charm.unit].update(
+                {"keystore-password": generate_password()}
+            )
+        if not self.truststore_password:
+            self.peer_relation.data[self.charm.unit].update(
+                {"truststore-password": generate_password()}
+            )
+
+        self._request_certificate()
+
+    def _on_certificate_available(self, event):
+        """"""
+        if not self.peer_relation:
+            event.defer()
+            return
+
+        # avoid setting tls files and restarting
+        if event.certificate_signing_request != self.csr:
+            logger.error("Can't use certificate, found unknown CSR")
+            return
+
+        self.peer_relation.data[self.charm.unit].update(
+            {"certificate": event.certificate, "ca": event.ca}
+        )
+
+        self.set_server_key()
+        self.set_ca()
+        self.set_certificate()
+        self.set_truststore()
+        self.set_keystore()
+
+    def _set_tls_private_key(self, event: ActionEvent) -> None:
+        """Handler for `set_tls_private_key` action."""
+        private_key = self._parse_tls_file(event.params.get("internal-key", None))
+        self.peer_relation.data[self.charm.unit].update({"private-key": private_key})
+
+        self._on_certificate_expiring(event)
+
+    @property
+    def peer_relation(self) -> Relation:
+        """Get the peer relation of the charm."""
+        return self.charm.peer_relation
+
+    @property
+    def enabled(self) -> bool:
+        """Flag to check if the cluster should run with TLS.
+
+        Returns:
+            True if TLS encryption should be active. Otherwise False
+        """
+        enabled = False
+        try:
+            enabled = self.peer_relation.data[self.charm.app].get("tls", None) == "enabled"
+        except AttributeError:
+            logger.info("Peer relation not initialized")
```

Comment:
If I'm right in assuming that this flag is to ensure that the peer relation exists first, I'd change to have a bare `if not self.charm.cluster.relation: logger.info("Peer relation not initialized") return False`

Reply:
Though I don't think this is necessary, we can just ensure that any event that requires peer relation data, has a check for it existing, else defer/return.

Reply:
so this would look something like `return self.peer_relation.data[self.charm.app].get('tls', None) == "enabled"`

Reply:
`KafkaAuth` will need to know if tls is enabled or not. This information will come from this method. On `charm.py`'s `__init__()`, when instantiating `KafkaProvider`, `KafkaAuth` is called.
On `cham.py` init, relation doesn't exist yet. I'm guessing this how `ops` handles the instantiation of things every time the python code on the charm is executed. So in this case, we need to take into account this edge case I think.

Reply:
The nice thing about properties though, is that they don't evaluate until they're called. So you can just ensure that whichever events you're calling 'tls.enabled', you have a minor check above that handled event for `if not self.cluster.relation: return`. It's clunky, but I see there being benefits of allowing library code (in this case tls.py) be agnostic to charm event wrangling, which can be managed by the events handlers wherever.

Reply:
I think that making the change so `KafkaAuth` solves this issue. As you said, since properties are not evaluated until called, now there is no issue during charm init :)